File: mf2/java/src/main/java/com/box/l10n/mojito/mf2/PluralRules.java
package com.box.l10n.mojito.mf2;

final class PluralRules {
    private PluralRules() {}

    static String selectCardinalPluralCategory(String locale, Object value) {
        return selectPluralCategory(locale, value, NumberSelect.PLURAL);
    }

    static String selectPluralCategory(String locale, Object value, NumberSelect numberSelect) {
        if (value == null || value instanceof Boolean || numberSelect == NumberSelect.EXACT) {
            return null;
        }

        CldrPluralRules.NumberOperands operands =
                CldrPluralRules.NumberOperands.fromString(Mf2Formatter.valueToString(value));
        if (operands == null) {
            return null;
        }

        return switch (numberSelect) {
            case PLURAL -> CldrPluralRules.selectCardinal(locale, operands);
            case ORDINAL -> CldrPluralRules.selectOrdinal(locale, operands);
            case EXACT -> null;
        };
    }
}

File: mf2/kotlin/src/main/kotlin/com/box/l10n/mojito/mf2/PluralRules.kt
package com.box.l10n.mojito.mf2

internal object PluralRules {
    private fun supportedOperands(value: String): NumberOperands? {
        val number = value.toDoubleOrNull() ?: return null
        if (!number.isFinite() || kotlin.math.abs(number) >= 9223372036854775808.0) return null
        return NumberOperands.fromString(value)
    }

    fun selectCardinal(locale: String, value: String): String =
        supportedOperands(value)?.let { CldrPluralRules.selectCardinal(locale, it) } ?: throw Mf2Error.badSelector("Numeric plural operand is outside the supported range.")

    fun selectOrdinal(locale: String, value: String): String =
        supportedOperands(value)?.let { CldrPluralRules.selectOrdinal(locale, it) } ?: throw Mf2Error.badSelector("Numeric plural operand is outside the supported range.")
}
